docs/guide/slots.md

```diff
@@ -282,6 +282,92 @@ Slot content can also be set using `#with_content`:
 <% end %>
 ```
 
+## Collection counter
+
+ViewComponent provides a counter parameter matching the component name followed by `_counter`.
+
+To access the counter, add it to `initialize` as an argument:
+
+```ruby
+# app/components/child_component.rb
+class ChildComponent < ViewComponent::Base
+  def initialize(first_name:, child_counter:)
+    @first_name, @child_counter = first_name, child_counter
+  end
+end
+```
+
+```erb
+<%# app/components/child_component.html.erb %>
+<li><%= @child_counter %>: <%= @first_name %></li>
+```
+
+```ruby
+# app/components/parent_component.rb
+class ParentComponent < ViewComponent::Base
+  renders_many :children, ChildComponent
+end
+```
+
+```erb
+<%# app/components/parent_component.html.erb %>
+<% children.each do |child| %>
+  <%= child %>
+<% end %>
+```
+
+```erb
+<%# index.html.erb %>
+<%= render ParentComponent.new do |component| %>
+  <% component.with_child(first_name: "Daniel") %>
+  <% component.with_child(first_name: "Peter") %>
+<% end %>
+```
+
+## Collection iteration context
+
+ViewComponent provides an iteration parameter matching the component name followed by `_iteration`. This gives contextual information about the iteration to components within the collection (`#size`, `#index`, `#first?`, and `#last?`).
+
+To access the iteration, add it to `initialize` as an argument:
+
+```ruby
+# app/components/child_component.rb
+class ChildComponent < ViewComponent::Base
+  def initialize(first_name:, child_iteration:)
+    @first_name, @iteration = first_name, child_iteration
+  end
+end
+```
+
+```erb
+<%# app/components/child_component.html.erb %>
+<li class="<%= "first_born" if @iteration.first? %>">
+  <%= @first_name %>
+</li>
+```
+
+```ruby
+#app/components/parent_component.rb
+class ParentComponent < ViewComponent::Base
+  renders_many :children, ChildComponent
+end
+```
+
+```erb
+#app/components/parent_component.html.erb
+<% children.each do |child| %>
```

Comment:
What happens if I call `sort`, or `reverse`? Will the iterator methods still work as-intended, or will they return misleading or incorrect values?

Reply:
@BlakeWilliams the order of the iterator is dependent on the order on which they are registered not the order in which they are sorted into inside the component, and while this may seem misleading I believe it may make more sense for the iteration context to be consistent with the order in which the slot setter is invoked

Reply:
I think the big thing for me here is that we can't reliably know _what_ the order is here. We're declaring it as first defined, but I imagine it would be easy for folks to assume that it would be based on rendering order.

e.g. any of the following rendering scenarios could lead to some unfortunate, and subtle bugs:

```ruby
children.each { |child| child.to_s }
children.reverse.each { |child| child.to_s }
children.sort.each { |child| child.to_s }
```

Since it's Ruby, it's also possible to do things like:

```ruby
def children
  super.reverse
end
```

In any of those cases the iterator methods we provide would return an incorrect value because of the lack of abstractions around rendering a collection of slots, and I don't believe that order of registration is a good substitute for that scenario.

So my general concern is that there's a good few ways to have drift between the reality of the iterator methods (first, last, etc) and what users expect to get back from the API. View Component leans heavily into "it's just Ruby" and I while I appreciate all the effort that went into this PR, I think that may be what we want to encourage folks to do here instead of providing an API on-top of iterators based on registration order for them to utilize.

Reply:
This is a really great point and you've changed my mind on this PR. Generally I care about the rendering order and not the order slots were added, at least in cases like I describe in the https://github.com/ViewComponent/view_component/issues/1737.